tpie/pipelining/virtual.h

```diff
@@ -671,18 +1119,249 @@ class vpush_node : public node {
+	src_t src;
+};
+
+template <typename Input, typename Output>
+class devirtualization_pull_factory: public factory_base {
+public:
+	template <typename src_t>
+	struct constructed {
+		using type = bits::devirtualize_pull_begin_node<Output>;
+	};
+
+	devirtualization_pull_factory(const devirtualization_pull_factory &) = delete;
+	devirtualization_pull_factory(devirtualization_pull_factory &&) = default;
+	devirtualization_pull_factory & operator=(const devirtualization_pull_factory &) = delete;
+	devirtualization_pull_factory & operator=(devirtualization_pull_factory &&) = default;
+
+	explicit devirtualization_pull_factory(virtual_chunk_pull<Input, Output> chunk) : chunk(chunk) {}
+
+	template <typename src_t>
+	bits::devirtualize_pull_begin_node<Input> construct(src_t && src) {
+		node_token tok = src.get_token();
+		auto srcnode = std::make_unique<bits::devirtualize_pull_end_node<src_t, Input>>(std::move(src), chunk);
+		this->init_sub_node(*srcnode);
+		bits::devirtualize_pull_begin_node<Output> r(chunk, std::move(srcnode));
+		this->init_sub_node(r);
+		return r;
+	}
+	
+	virtual_chunk_pull<Input, Output> chunk;
+};
+
+	
+	
 } // namespace bits
 
+/**
+ * \brief Convert a virtual_chunk_end for use at the end of a normal none virtual pipeline
+ */
+template <typename Input>
+pipe_end<termfactory<bits::devirtualize_end_node<Input>, virtual_chunk_end<Input>>> devirtualize(const virtual_chunk_end<Input> & out) {
+	return out;
+}
+
+/**
+ * \brief Convert a virtual_chunk_begin for use at the beginning of a normal none virtual pipeline
+ */
+template <typename Output>
+pipe_begin<tfactory<bits::devirtualize_begin_node, Args<Output>, virtual_chunk_begin<Output>>> devirtualize(const virtual_chunk_begin<Output> & in) {
+	return in;
+}
+
+/**
+ * \brief Convert a virtual_chunk for use in the middle of a normal none virtual pipeline
+ */
+template <typename Input, typename Output>
+pipe_middle<bits::devirtualization_factory<Input, Output>> devirtualize(const virtual_chunk<Input, Output> & mid) {
+	return {mid};
+}
+
+/**
+ * \brief Convert a virtual_chunk_pull_begin for use at the beginning of a normal none virtual pull pipeline
+ */
+template <typename Input>
+pullpipe_begin<termfactory<bits::devirtualize_pull_begin_node<Input>, virtual_chunk_pull_begin<Input>>> devirtualize(const virtual_chunk_pull_begin<Input> & out) {
+	return out;
+}
+
+/*
+ * \brief Convert a virtual_chunk_pull_end for use at the end of a normal none virtual pull pipeline
+ */
+template <typename Input>
+pullpipe_end<tfactory<bits::devirtualize_pull_end_node, Args<Input>, virtual_chunk_pull_end<Input>>> devirtualize(const virtual_chunk_pull_end<Input> & in) {
+	return in;
+}
+
+/*
+ * \brief Convert a virtual_chunk_pull for use in the middle of a normal none virtual pull pipeline
+ */
+template <typename Input, typename Output>
+pullpipe_middle<bits::devirtualization_pull_factory<Input, Output>> devirtualize(const virtual_chunk_pull<Input, Output> & mid) {
+	return {mid};
```

Comment:
Maybe use explicit constructors `{...}` for all these six devirtualize overloads?